File: Java/ProjetoJavaTimSort/src/main/java/com/timsort_java/benchmarks/MainBenchmark.java
package com.timsort_java.benchmarks;

import com.timsort_java.structures.ArrayList;
import com.timsort_java.structures.ArrayQueue;
import com.timsort_java.structures.ArrayStack;
import com.timsort_java.structures.LinkedList;
import com.timsort_java.structures.LinkedQueue;
import com.timsort_java.structures.LinkedStack;
import com.timsort_java.timsort_logic.TimSort;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Consumer;

public class MainBenchmark {

    public static void runBenchmark(String dataFilepath, String outputFilePath) {
        int[] data = BenchmarkUtils.loadDataFromCsv(dataFilepath);

        Map<String, Object> structures = new LinkedHashMap<>();
        structures.put("ArrayList", new ArrayList());
        structures.put("ArrayQueue", new ArrayQueue());
        structures.put("ArrayStack", new ArrayStack());
        structures.put("LinkedList", new LinkedList());
        structures.put("LinkedQueue", new LinkedQueue());
        structures.put("LinkedStack", new LinkedStack());

        // Clear the output file for each new benchmark run (as in Python, it appends a new table)
        // This is a slight deviation from Python's 'a' mode, but necessary to prevent duplicate headers
        // and ensure each run starts fresh for its set of results.
        // The Python script adds a newline, effectively starting a new table for each run.
        // Here, we'll append a newline before writing the new set of results if the file is not empty.
        try (PrintWriter writer = new PrintWriter(new FileWriter(outputFilePath, true))) {
            if (new File(outputFilePath).length() > 0) {
                writer.println(); // Add a newline if file is not empty
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        for (Map.Entry<String, Object> entry : structures.entrySet()) {
            String name = entry.getKey();
            Object structure = entry.getValue();

            System.out.println(String.format("Running benchmark for %s with %d elements...", name, data.length));

            // Copy data to a fresh structure for each benchmark run
            BenchmarkUtils.copyArrayToStructure(data, structure);

            // Create a Consumer for the TimSort function
            Consumer<int[]> timsortConsumer = arr -> TimSort.timsort(arr);

            double[] performance = BenchmarkUtils.measurePerformance(timsortConsumer, structure);
            double timeTaken = performance[0];
            double memoryUsed = performance[1];

            BenchmarkUtils.writeResultsToCsv(outputFilePath, name, data.length, timeTaken, memoryUsed);
            System.out.println(String.format("  Time: %.6f seconds, Memory: %.4f MB", timeTaken, memoryUsed));
        }

        System.out.println(String.format("Benchmark results saved to %s", outputFilePath));
    }
}

File: Python/benchmarks/main_benchmark.py
from timsort.timsort import timsort
from structures.array_based import ArrayList, ArrayQueue, ArrayStack
from structures.linked_based import LinkedList, LinkedQueue, LinkedStack
from benchmarks.benchmark_utils import load_data_from_csv, copy_array_to_structure, measure_performance
import os
import csv

def run_benchmark(data_filepath, output_filepath):
    data = load_data_from_csv(data_filepath)
    results = []

    structures = {
        "ArrayList": ArrayList(),
        "ArrayQueue": ArrayQueue(),
        "ArrayStack": ArrayStack(),
        "LinkedList": LinkedList(),
        "LinkedQueue": LinkedQueue(),
        "LinkedStack": LinkedStack()
    }

    for name, structure in structures.items():
        print(f"Running benchmark for {name} with {len(data)} elements...")
        copy_array_to_structure(data, structure)
        time_taken, memory_used = measure_performance(timsort, structure)
        results.append({
            "structure": name,
            "data_size": len(data),
            "time_taken": time_taken,
            "memory_used_mb": memory_used
        })
        print(f"  Time: {time_taken:.6f} seconds, Memory: {memory_used:.4f} MB")

    # Append results to CSV
    with open(output_filepath, 'a', newline='') as csvfile:
        fieldnames = ["structure", "data_size", "time_taken", "memory_used_mb"]
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

        if os.path.getsize(output_filepath) == 0:
            writer.writeheader()

        csvfile.write("\n")  # Add an empty line between each table

        for row in results:
            writer.writerow(row)

    print(f"Benchmark results saved to {output_filepath}")


